apps/sms/services/js/moz_mobile_message/moz_mobile_message_shim.js

```diff
@@ -0,0 +1,231 @@
+/* vim: set shiftwidth=2 tabstop=2 autoindent cindent expandtab: */
+
+
+/* global bridge */
+
+'use strict';
+(function() {
+
+/**
+ * Name of the service for mozMobileMessage API shim
+ * @type {string}
+ */
+
+const SERVICE_NAME = 'mozMobileMessageShim';
+
+/**
+ * Array of the event name that is corresponding to mozMobileMessage events.
+ * @type {Array}
+ */
+
+const EVENTS = ['received', 'sending', 'sent', 'failed', 'deleted',
+                'readsuccess', 'deliverysuccess'];
+
+/**
+ * Array of method that need to be exposed for API shim.
+ * @type {Array}
+ */
+
+const METHODS = ['getMessage', 'retrieveMMS', 'send', 'sendMMS',
+                 'delete', 'markMessageRead', 'getSegmentInfoForText'];
+
+/**
+ * Array of stream that need to return data(messages/threads) in chunk.
+ * @type {Array}
+ */
+
+const STREAMS = ['getThreads', 'getMessages'];
+
+var mozMobileMessage = navigator.mozMobileMessage;
+var service = bridge.service(SERVICE_NAME);
+
+var MozMobileMessageShim = {
+  init: function ms_init() {
+    EVENTS.forEach((event) => {
+      mozMobileMessage.addEventListener(
+        event,
+        this[Utils.camelCase(`on-${event}`)]
+      );
+    });
+
+    METHODS.forEach((shimMethod) => {
+      service.method(shimMethod, this[shimMethod]);
+    });
+
+    STREAMS.forEach((shimStream) => {
+      service.stream(shimStream, this[shimStream]);
+    });
+  },
+
+  /* Events */
+
+  onSending: function ms_onSending(e) {
+    service.broadcast('message-sending', { message: e.message });
+  },
+
+  onFailed: function ms_onFailed(e) {
+    service.broadcast('message-failed-to-send', { message: e.message });
+  },
+
+  onDeliverysuccess: function ms_onDeliverysuccess(e) {
+    service.broadcast('message-delivered', { message: e.message });
+  },
+
+  onReadsuccess: function ms_onReadsuccess(e) {
+    service.broadcast('message-read', { message: e.message });
+  },
+
+  onSent: function ms_onSent(e) {
+    service.broadcast('message-sent', { message: e.message });
+  },
+
+  onReceived: function ms_onReceived(e) {
+    var message = e.message;
+
+    if (message.messageClass && message.messageClass === 'class-0') {
+      return;
+    }
+
+    // Here we can only have one sender, so deliveryInfo[0].deliveryStatus =>
+    // message status from sender. Ignore 'pending' messages that are received
+    // this means we are in automatic download mode
+    if (message.delivery === 'not-downloaded' &&
+        message.deliveryInfo[0].deliveryStatus === 'pending') {
+      return;
+    }
+
+    service.broadcast('message-received', { message: message });
+  },
+
+  onDeleted: function ms_onDeleted(e) {
+    if (e.deletedThreadIds && e.deletedThreadIds.length) {
+      service.broadcast('threads-deleted', {
+        ids: e.deletedThreadIds
+      });
+    }
+  },
+
+  /* Method */  
+
+  getMessage: function ms_getMessag(id) {
+    return mozMobileMessage.getMessage(id);
```

Comment:
I still split the method into identical function here instead of `service.method(shimMethod, mozMobileMessage[shimMethod])` because I don't think all the request result could be passed directly(at least getMessage will return error with error: The object could not be cloned). So we might need some clone for some case to return data properly.